src/wallet/message_store.cpp

```diff
@@ -0,0 +1,1389 @@
+            data.receiving_member_index = j;
+            data_list.push_back(data);
+          }
+          return true;
+        }
+        else
+        {
+          // Somebody else sent this to us: We can sign it
+          // It would be possible to just pass it on, but that's not directly supported here
+          processing_data data;
+          data.processing = message_processing::sign_tx;
+          data.message_ids.push_back(m.id);
+          data_list.push_back(data);
+          return true;
+        }
+      }
+
+      case message_type::note:
+        note_found = true;
+        break;
+
+      case message_type::multisig_sync_data:
+        sync_data_found = true;
+        break;
+
+      default:
+        break;
+      }
+    }
+  }
+  if (waiting_found)
+  {
+    wait_reason = tr("There are waiting messages, but nothing ready to process under normal circumstances");
+    if (sync_data_found)
+    {
+      wait_reason += tr("\nUse \"mms next sync\" if you want to force processing of the waiting sync data");
+    }
+    if (note_found)
+    {
+      wait_reason += tr("\nUse \"mms note\" to display the waiting notes");
+    }
+  }
+  else
+  {
+    wait_reason = tr("There are no messages waiting to be processed.");
+  }
+
+  return false;
+}
+
+void message_store::set_messages_processed(const processing_data &data)
+{
+  for (size_t i = 0; i < data.message_ids.size(); ++i)
+  {
+    set_message_processed_or_sent(data.message_ids[i]);
+  }
+}
+
+void message_store::set_message_processed_or_sent(uint32_t id)
+{
+  uint32_t index = get_message_index_by_id(id);
+  message &m = m_messages[index];
+  if (m.state == message_state::waiting)
+  {
+    // So far a fairly cautious and conservative strategy: Only delete from Bitmessage
+    // when fully processed (and e.g. not already after reception and writing into
+    // the message store file)
+    delete_transport_message(id);
+    m.state = message_state::processed;
+  }
+  else if (m.state == message_state::ready_to_send)
+  {
+    m.state = message_state::sent;
+  }
+  m.modified = time(NULL);
+}
+
+void message_store::encrypt(crypto::public_key public_key, const std::string &plaintext,
+                            std::string &ciphertext, crypto::public_key &encryption_public_key, crypto::chacha_iv &iv)
+{
+  crypto::secret_key encryption_secret_key;
+  crypto::generate_keys(encryption_public_key, encryption_secret_key);
+
+  crypto::key_derivation derivation;
+  bool success = crypto::generate_key_derivation(public_key, encryption_secret_key, derivation);
+  THROW_WALLET_EXCEPTION_IF(!success, tools::error::wallet_internal_error, "Failed to generate key derivation for message encryption");
+
+  crypto::chacha_key chacha_key;
+  crypto::generate_chacha_key(&derivation, sizeof(derivation), chacha_key, 1);
+  iv = crypto::rand<crypto::chacha_iv>();
+  ciphertext.resize(plaintext.size());
+  crypto::chacha20(plaintext.data(), plaintext.size(), chacha_key, iv, &ciphertext[0]);
+}
```

Comment:
As I understand it, the IRC discussion with Sarang last Sunday showed that we are in the clear here with "chacha" being a stream cipher able to deal with arbitrary messages sizes, with using a new IV value for each message to avoid key reuse, and with authenticating messages by using signatures.

I did not change anything here.

Reply:
The point had nothing to do with message sizes fwiw, just key derivation.

Reply:
Right. Resolved.